File: app/src/main/java/com/chuongvd/app/signal/data/source/remote/service/rx/RocketSubscriber.java
package com.chuongvd.app.signal.data.source.remote.service.rx;

import android.support.annotation.NonNull;
import com.chuongvd.app.signal.data.source.remote.service.BaseException;
import io.reactivex.observers.DisposableObserver;

public abstract class RocketSubscriber<T> extends DisposableObserver<T> {

    private T data;

    @Override
    protected void onStart() {
        super.onStart();
    }

    @Override
    public void onNext(@NonNull T t) {
        data = t;
    }

    @Override
    public void onError(@NonNull Throwable e) {
        BaseException exception;
        if (e instanceof BaseException) {
            exception = (BaseException) e;
        } else {
            exception = BaseException.toUnexpectedError(e);
        }

        onRequestFinish();
        onError(exception);
    }

    @Override
    public void onComplete() {
        onSuccess(data);
        onRequestFinish();
    }

    public abstract void onSuccess(T object);

    public abstract void onError(BaseException error);

    /**
     * Runs after request complete or error
     **/
    public void onRequestFinish() {

    }
}

File: app/src/main/java/com/chuongvd/app/signal/data/source/remote/service/rx/RocketSubscriber.kt
package com.chuongvd.app.signal.data.source.remote.service.rx

import com.chuongvd.app.signal.data.source.remote.service.BaseException
import io.reactivex.observers.DisposableObserver

abstract class RocketSubscriber<T> : DisposableObserver<T>() {

  private var data: T? = null

  override fun onStart() {
    super.onStart()
  }

  override fun onNext(t: T) {
    data = t
  }

  override fun onError(e: Throwable) {
    val exception: BaseException
    if (e is BaseException) {
      exception = e
    } else {
      exception = BaseException.toUnexpectedError(e)
    }

    onRequestFinish()
    onError(exception)
  }

  override fun onComplete() {
    onSuccess(data)
    onRequestFinish()
  }

  abstract fun onSuccess(`object`: T?)

  abstract fun onError(error: BaseException)

  /**
   * Runs after request complete or error
   */
  fun onRequestFinish() {

  }
}